### PR TITLE
luigidelrosario-scrapy_upcat_scraper

### PR DESCRIPTION
**Pull request by luigidelrosario**
The code scrapes for the list of UPCAT passers from https://upcat.upd.edu.ph/results and outputs it to a text file. The downloader middleware ignores all requests after 5 successful responses.

Notes: the output file is rewritten everytime you run the scraper.

### CODE REVIEW

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+"""EXERCISE 3: UPCAT Results Spider using Scrapy."""
+
+import scrapy
+from bs4 import BeautifulSoup
+from scrapy.linkextractors import LinkExtractor
+
+
+class UPCATPasser(scrapy.Item):
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: Move this over to `items.py`

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/pipelines.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class DelrosarioUpcatResultsPipeline(object):
+	"""Item pipeline class."""
+	
+    def process_item(self, item, spider):
+    	f = codecs.open('upcat_passers.txt', 'a+', 'utf-8')
+        f.write(item['name'] + '\n')
+        f.write(item['campus'] + '\n')
+        f.write(item['course'] + '\n\n')
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/pipelines.py]: You can simplify these into: `f.write('{name}\n{campus}\n{course}'.format(**item))`

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/middlewares.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    def process_request(self, request, spider):
+        """Ignore everything after getting 5 requests."""
+        try:
+            if (request.meta['count'] > 5):
+                raise IgnoreRequest()
+            else:
+                return None
+        except KeyError:
+            return None
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/middlewares.py]: `request.meta.get('count', DEFAULT_VALUE)` so you don't need to worry about key errors

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/middlewares.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+class DelrosarioUpcatResultsDownloaderMiddleware(object):
+    """Downloader middleware that stops after 5 requests."""
+
+    def process_request(self, request, spider):
+        """Ignore everything after getting 5 requests."""
+        try:
+            if (request.meta['count'] > 5):
+                raise IgnoreRequest()
+            else:
+                return None
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/middlewares.py]: No need for this

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            meta={'count': 0}
+        )
+
+    def get_next_pages(self, response):
+        """Use LinkExtractor to get next pages."""
+        link_extractor = LinkExtractor(allow=r'.*page-\d+.html')
+        links = link_extractor.extract_links(response)
+        count = 0
+        for link in links:
+            count += 1
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: look at `enumerate` so you don't have to manually count everything

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    """Class for spider."""
+
+    name = 'upcat_results_spider'
+    url = 'https://upcat.up.edu.ph/results/'
+
+    def start_requests(self):
+        """Reset the file, and get the index page."""
+        f = open('upcat_passers.txt', 'r+')
+        f.truncate()
+        f.close()
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: Why are you doing this?

Comment (luigidelrosario) [reply scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: The item pipeline appends to the output file. This ensures that whenever the scraper is run, the file is blank.

Comment (naromsantos) [reply scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: It's okay if it appends it for now, no need to clear it every time :+1:

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/items.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+
+"""Item template for UPCAT Passer."""
+
+import scrapy
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/items.py]: `from scrapy import Item`

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+"""EXERCISE 3: UPCAT Results Spider using Scrapy (main file)."""
+
+import scrapy
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: `from scrapy import Spider, Request`

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+        for passer_entry in soup.find_all('tr'):
+
+            details = passer_entry.find_all(
+                'span',
+                {'style': 'font-size:small;'}
+            )
+
+            if details:
+                upcat_passer = UPCATPasserItem(
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: `yield UPCATPasserItem()`

File: scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            callback=self.get_next_pages,
+            meta={'count': 0}
+        )
+
+    def get_next_pages(self, response):
+        """Use LinkExtractor to get next pages."""
+        link_extractor = LinkExtractor(allow=r'.*page-\d+.html')
+        links = link_extractor.extract_links(response)
+        for count, link in enumerate(links):
+            yield scrapy.Request(
```
Comment (naromsantos) [inline scrapers/delrosario_upcat_results/delrosario_upcat_results/spiders/upcat_spider.py]: `response.follow(link, callback, meta)`

**Pull request merged**